pkg/util/watch/delete.go

```diff
@@ -0,0 +1,186 @@
+		// NewIndexerInformer returns cache.cache as the indexer that implements
+		// the cache.Store, but the cache is populated by DeltaFIFO, which
+		// wraps objects with DeletedFinalStateUnknown sometimes.
+		staleObj, stale := lastKnownObj.(cache.DeletedFinalStateUnknown)
+		if stale {
+			// Object was deleted but the watch deletion event was missed while
+			// disconnected from apiserver. So the last known object may be
+			// stale, but metadata.deletionTimestamp may still be available, if
+			// that update was received before the disconnect.
+			lastKnownObj = staleObj.Obj
+		}
+		if lastKnownObj == nil {
+			// The object either never existed or was deleted before the initial
+			// List or both created and deleted after List but while Watch was
+			// disconnected from the apiserver.
+			// Alternatively, the cache/Indexer.Get may have errored
+			// (errors are swallowed by DeltaFIFO.Replace).
+
+			// Stop waiting - object already deleted or never existed
+			klog.V(3).Infof("Object already deleted %q=%q, %q=%q",
+				logFieldObjectRef, id.ObjectKey.String(),
+				logFieldObjectKind, id.Kind)
+			return true, nil
+		}
+		// Since we supplied a client.Object, we know the objects returned
+		// will be client.Object.
+		cObj, ok := lastKnownObj.(client.Object)
+		if !ok {
+			return false, errors.Errorf("unexpected watch cache object %T for object %s %s: %+v",
+				lastKnownObj, id.Kind, id.ObjectKey, lastKnownObj)
+		}
+		// Check if the object was already deleted but still terminating
+		if cObj.GetDeletionTimestamp() != nil {
+			// Continue waiting until fully deleted - already deleting
+			klog.V(3).Infof("Object already deleting %q=%q, %q=%q",
+				logFieldObjectRef, id.ObjectKey.String(),
+				logFieldObjectKind, id.Kind)
+			// Note: Whoever initiated the delete chose what deletion
+			// propagation policy to use. We can't change that now.
+			// If Foreground was used, the metav1.FinalizerDeleteDependents
+			// finalizer will have been added to enforce it.
+			return false, nil
+		}
+
+		klog.V(1).Infof("Deleting object %q=%q, %q=%q",
+			logFieldObjectRef, id.ObjectKey.String(),
+			logFieldObjectKind, id.Kind)
+
+		// Object exists - safe to delete now
+		// Use foreground deletion to ensure all managed objects are garbage
+		// collected by the apiserver before continuing.
+		if err := c.Delete(ctx, cObj, client.PropagationPolicy(metav1.DeletePropagationForeground)); err != nil {
+			if apierrors.IsNotFound(err) {
+				klog.V(3).Infof("Object already deleted %q=%q, %q=%q",
+					logFieldObjectRef, id.ObjectKey.String(),
+					logFieldObjectKind, id.Kind)
+				// Stop waiting - object already deleted
+				return true, nil
+			}
+			// Stop waiting
+			return false, err
+		}
+		klog.V(3).Infof("Object delete successful %q=%q, %q=%q",
+			logFieldObjectRef, id.ObjectKey.String(),
+			logFieldObjectKind, id.Kind)
+
+		// Continue waiting until fully deleted
+		return false, nil
```

Comment:
Returning false tells UntilDeletedWithSync to keep processing events. In this case, it hasn't received/processed the Deleted event yet. When it does, it will stop and return.